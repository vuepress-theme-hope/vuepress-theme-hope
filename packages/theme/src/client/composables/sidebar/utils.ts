import { h } from "vue";
import { useRoute } from "vue-router";
import { useIconPrefix } from "../themeData";
import AutoLink from "../../components/AutoLink";
import { isActiveLink } from "../../utils";

import type { VNode } from "vue";
import type {
  ResolvedSidebarItem,
  ResolvedSidebarHeaderItem,
  ResolvedSidebarGroupItem,
} from "../../../shared";

export const renderIcon = (icon?: string): VNode | null =>
  icon
    ? h("i", {
        class: ["iconfont", `${useIconPrefix().value}${icon}`],
      })
    : null;

export const renderItem = (
  config: ResolvedSidebarItem,
  props: VNode["props"]
): VNode => {
  return config.link
    ? // if the item has link, render it as `<AutoLink>`
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore
      h(AutoLink, {
        ...props,
        config,
      })
    : // if the item only has text, render it as `<p>`
      h("p", props, [renderIcon(config.icon), config.text]);
};

export const renderGroupHeader = (
  { collapsable, icon, text }: ResolvedSidebarGroupItem,
  open: boolean
): (VNode | null)[] => [
  // icon
  renderIcon(icon),
  // title
  h("span", { class: "title" }, text),
  // arrow
  collapsable
    ? h("span", { class: ["arrow", open ? "down" : "right"] }, text)
    : null,
];

export const renderChildren = (
  children: ResolvedSidebarHeaderItem[]
): VNode | null => {
  const route = useRoute();
  if (!children) return null;

  return h(
    "ul",
    { class: "sidebar-sub-headers" },
    children.map((child) => {
      const active = isActiveLink(route, child.link);

      return h("li", { class: "sidebar-sub-header" }, [
        renderItem(child, {
          class: {
            "sidebar-link": true,
            heading: true,
            active,
          },
        }),
        renderChildren(child.children),
      ]);
    })
  );
};
