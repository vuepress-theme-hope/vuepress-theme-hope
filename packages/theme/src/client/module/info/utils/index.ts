export * from "./define";
export * from "./editlink";
