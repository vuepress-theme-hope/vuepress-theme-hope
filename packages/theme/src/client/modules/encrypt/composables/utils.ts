import { type ComputedRef, computed } from "vue";

import { useThemeData } from "@theme-hope/composables/index";

import { type EncryptConfig } from "../../../../shared/index.js";

export const useEncryptData = (): ComputedRef<EncryptConfig> => {
  const themeData = useThemeData();

  return computed(() => themeData.value.encrypt || {});
};
