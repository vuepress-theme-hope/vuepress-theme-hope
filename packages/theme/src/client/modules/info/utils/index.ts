export * from "./editLink.js";
export * from "./typings.js";
