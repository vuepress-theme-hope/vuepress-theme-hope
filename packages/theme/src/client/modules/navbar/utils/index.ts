import {
  type AutoLinkOptions,
  type NavGroup,
  type NavbarItem,
} from "../../../../shared/index.js";

// resolved
export type ResolvedThemeNavbarItem =
  | NavbarItem
  | NavGroup<AutoLinkOptions | NavGroup<AutoLinkOptions>>;
