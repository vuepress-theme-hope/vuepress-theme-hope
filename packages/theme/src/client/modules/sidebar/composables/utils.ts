import { ensureEndingSlash } from "@vuepress/shared";

export const resolvePrefix = (prefix = "", path = ""): string =>
  path.startsWith("/") ? path : `${ensureEndingSlash(prefix)}${path}`;
