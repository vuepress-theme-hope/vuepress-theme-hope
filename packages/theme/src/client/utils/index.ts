export * from "./getAncestorLinks.js";
export * from "./resolveLinkInfo.js";
