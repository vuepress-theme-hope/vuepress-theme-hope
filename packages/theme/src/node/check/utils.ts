export const PLUGIN_CHECKLIST = [
  ["@vuepress/plugin-active-header-links", "activeHeaderLinks"],
  ["@vuepress/plugin-blog", "blog"],
  ["@vuepress/plugin-catalog", "catalog"],
  ["@vuepress/plugin-comment", "comment"],
  ["@vuepress/plugin-copy-code", "copyCode"],
  ["@vuepress/plugin-copyright", "copyright"],
  ["@vuepress/plugin-docsearch", "docsearch"],
  ["@vuepress/plugin-external-link-icon", "externalLinkIcon"],
  ["@vuepress/plugin-feed", "feed"],
  ["@vuepress/plugin-git", "git"],
  ["@vuepress/plugin-links-check", "linksCheck"],
  ["@vuepress/plugin-notice", "notice"],
  ["@vuepress/plugin-nprogress", "nprogress"],
  ["@vuepress/plugin-photo-swipe", "photoSwipe"],
  ["@vuepress/plugin-prismjs", "prismjs"],
  ["@vuepress/plugin-pwa", "pwa"],
  ["@vuepress/plugin-reading-time", "readingTime"],
  ["@vuepress/plugin-redirect", "redirect"],
  ["@vuepress/plugin-rtl", "", 'Set "rtl: true" in the needed theme locales.'],
  ["@vuepress/plugin-search", "search"],
  ["@vuepress/plugin-seo", "seo"],
  ["@vuepress/plugin-sitemap", "sitemap"],
  ["@vuepress/plugin-theme-data", "", "This plugin is called internally."],
  ["@vuepress/plugin-watermark", "watermark"],
  ["vuepress-plugin-components", "components"],
  ["vuepress-plugin-md-enhance", "mdEnhance"],
  ["vuepress-plugin-search-pro", "searchPro"],
];
