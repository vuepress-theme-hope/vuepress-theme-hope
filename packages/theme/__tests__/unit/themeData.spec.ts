/* eslint-disable @typescript-eslint/naming-convention */
import { createBaseApp } from "@vuepress/core";
import { path } from "@vuepress/utils";
import { describe, expect, it } from "vitest";

import { emptyTheme } from "./__fixtures__/theme/empty.js";
import { getStatus, getThemeData } from "../../src/node/config/index.js";
import { type ThemeOptions } from "../../src/shared/index.js";

describe("should generate themeData correctly", () => {
  it("Should contain basic properties", () => {
    const app = createBaseApp({
      locales: {
        "/": {
          lang: "en-US",
        },
      },
      bundler: {} as any,
      source: path.resolve(__dirname, "./__fixtures__/src"),
      theme: emptyTheme,
    });
    const themeOptions = {};

    const result = getThemeData(app, themeOptions, getStatus(app, {}));

    expect(result).toHaveProperty("encrypt");
    expect(result.locales).toBeTypeOf("object");
    expect(Object.keys(result.locales)).toEqual(["/"]);
    expect(result).toMatchSnapshot();
  });

  it("Should handle single language", () => {
    const app = createBaseApp({
      locales: {
        "/": {
          lang: "en-US",
        },
      },
      bundler: {} as any,
      source: path.resolve(__dirname, "./__fixtures__/src"),
      theme: emptyTheme,
    });
    const themeOptions = {
      navbar: ["/", "/about"],
      sidebar: ["/", "/about"],
    };

    const result = getThemeData(app, themeOptions, getStatus(app, {}));

    expect(result.locales).toBeTypeOf("object");
    expect(Object.keys(result.locales)).toEqual(["/"]);
    expect(result.locales["/"].navbar).toEqual(["/", "/about"]);
    expect(result.locales["/"].sidebar).toEqual(["/", "/about"]);
  });

  it("locale should have higher property", () => {
    const app = createBaseApp({
      locales: {
        "/": {
          lang: "en-US",
        },
        "/zh/": {
          lang: "zh-CN",
        },
      },
      bundler: {} as any,
      source: path.resolve(__dirname, "./__fixtures__/src"),
      theme: emptyTheme,
    });
    const themeOptions = {
      navbar: ["/", "/about"],

      locales: {
        "/": {
          navbar: ["/", "/guide", "/about"],
        },
        "/zh/": {
          navbar: ["/zh/", "/zh/guide", "/zh/about"],
        },
      },
    };

    const result = getThemeData(app, themeOptions, getStatus(app, {}));

    expect(result.locales).toBeTypeOf("object");
    expect(new Set(Object.keys(result.locales))).toEqual(
      new Set(["/", "/zh/"])
    );
    expect(result.locales["/"].navbar).toEqual(["/", "/guide", "/about"]);
    expect(result.locales["/zh/"].navbar).toEqual([
      "/zh/",
      "/zh/guide",
      "/zh/about",
    ]);
  });

  it("should fallback to root if locale config is missing", () => {
    const app = createBaseApp({
      locales: {
        "/": {
          lang: "en-US",
        },
        "/zh/": {
          lang: "zh-CN",
        },
      },
      bundler: {} as any,
      source: path.resolve(__dirname, "./__fixtures__/src"),
      theme: emptyTheme,
    });
    const themeOptions = {
      navbar: ["/", "/guide", "/about"],

      locales: {
        "/": {},
        "/zh/": {
          navbar: ["/zh/", "/zh/guide", "/zh/about"],
        },
      },
    };

    const result = getThemeData(app, themeOptions, getStatus(app, {}));

    expect(result.locales["/"].navbar).toEqual(["/", "/guide", "/about"]);
    expect(result.locales["/zh/"].navbar).toEqual([
      "/zh/",
      "/zh/guide",
      "/zh/about",
    ]);
  });

  it("root only option should not appear in locales", () => {
    const app = createBaseApp({
      locales: {
        "/": {
          lang: "en-US",
        },
      },
      bundler: {} as any,
      source: path.resolve(__dirname, "./__fixtures__/src"),
      theme: emptyTheme,
    });
    const themeOptions: ThemeOptions = {
      pure: true,
      darkmode: "disable",
      encrypt: {},
    };

    const result = getThemeData(app, themeOptions, getStatus(app, {}));

    expect(result.locales["/"]).not.toHaveProperty("darkmode");
    expect(result.locales["/"]).not.toHaveProperty("encrypt");
    expect(result.locales["/"]).not.toHaveProperty("pure");
  });
});
