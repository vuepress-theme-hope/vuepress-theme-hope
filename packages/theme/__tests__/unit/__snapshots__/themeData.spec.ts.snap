// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should generate themeData correctly > Should contain basic properties 1`] = `
{
  "encrypt": {},
  "locales": {
    "/": {
      "encryptLocales": {
        "errorHint": "Please enter the correct password!",
        "iconLabel": "Page Encrypted",
        "placeholder": "Enter password",
        "remember": "Remember password",
      },
      "lang": "en-US",
      "metaLocales": {
        "author": "Author",
        "category": "Category",
        "contributors": "Contributors",
        "date": "Writing Date",
        "editLink": "Edit this page",
        "lastUpdated": "Last update",
        "next": "Next",
        "origin": "Original",
        "prev": "Prev",
        "print": "Print",
        "readingTime": "Reading Time",
        "tag": "Tag",
        "toc": "On This Page",
        "views": "Page views",
        "words": "Words",
      },
      "navbarLocales": {
        "langName": "English",
        "selectLangAriaLabel": "Select language",
      },
      "outlookLocales": {
        "darkmode": "Theme Mode",
        "fullscreen": "Full Screen",
        "themeColor": "Theme Color",
      },
      "routeLocales": {
        "back": "Go back",
        "home": "Take me home",
        "notFoundMsg": [
          "There’s nothing here.",
          "How did we get here?",
          "That’s a Four-Oh-Four.",
          "Looks like we've got some broken links.",
        ],
        "notFoundTitle": "Page not found",
        "openInNewWindow": "Open in new window",
        "skipToContent": "Skip to main content",
      },
    },
  },
}
`;
